core/src/main/scala/com/rallyhealth/vapors/core/data/NamedLens.scala

```diff
@@ -1,63 +1,242 @@
-      path = lens.path ::: this.path, // TODO: Why is this broken?
+      path = lens.path ::: this.path,
       get = get.andThen(lens.get),
     )
   }
 
+  // Helpful for building a tuple with the lens itself. Identical to identity, but you don't have to specify the type.
+  override def self: FieldLens[A, B] = this
+
+  override def paths: NonEmptyVector[DataPath] = NonEmptyVector.of(path)
+
   def field[C](
     name: String,
     getter: B => C,
-  ): NamedLens[A, C] = {
+  ): FieldLens[A, C] = {
     copy(
       path = path.atField(name),
       get = this.get.andThen(getter),
     )
   }
 
-  def atKey[C[x, y] <: Map[x, y], K : ValidDataPathKey, V](
+  // TODO: Use monoidK over some M here? maybe a separate method?
+  def atKey[K : ValidDataPathKey, V](
     key: K,
   )(implicit
-    ev: B <:< C[K, V],
-  ): NamedLens[A, Option[V]] = {
+    CI: Indexed[B, K, Option[V]],
+  ): FieldLens[A, Option[V]] = {
     copy(
       path = path.atKey(key),
-      get = get.andThen(b => ev(b).get(key)),
+      get = get.andThen(b => CI.get(b)(key)),
     )
   }
 
-  def head[C[x] <: Iterable[x], V](implicit ev: B <:< C[V]): NamedLens[A, Option[V]] = {
+  def filterKeys[K : ValidDataPathKey, V : Semigroup](
+    keys: NonEmptySet[K],
+  )(implicit
+    CI: Indexed[B, K, V],
+  ): FieldLens[A, V] = {
+    import syntax._
+    copy(
+      path = path.filterKeys(keys),
+      get = get.andThen(b => b.filterKeys(keys)),
+    )
+  }
+
+  def headOption[C[x] <: Iterable[x], V](implicit ev: B <:< C[V]): FieldLens[A, Option[V]] = {
     copy(
       path = path.atHead,
       get = get.andThen(b => ev(b).headOption),
     )
   }
+
+  def asIterable[V](implicit ev: B <:< IterableOnce[V]): AsIterableBuilder[V] = new AsIterableBuilder(ev)
+
+  // TODO: move to implicit class?
+  final class AsIterableBuilder[V](private val ev: B <:< IterableOnce[V]) {
+
+    def to[C](factory: Factory[V, C]): FieldLens[A, C] =
+      copy(get = outer.get.andThen(c => ev(c).iterator.to(factory)))
+  }
+
+  // TODO: Better way to abstract over arity?
+  def split[C, D](
+    bc: B ==> C,
+    bd: B ==> D,
+  ): CompoundLens[A, (C, D)] =
+    CompoundLens.of(
+      this.andThen(bc(NamedLens.id)),
+      this.andThen(bd(NamedLens.id)),
+    )
+
+  def split[C, D, E](
+    bc: B ==> C,
+    bd: B ==> D,
+    be: B ==> E,
+  ): CompoundLens[A, (C, D, E)] =
+    CompoundLens.of(
+      this.andThen(bc(NamedLens.id)),
+      this.andThen(bd(NamedLens.id)),
+      this.andThen(be(NamedLens.id)),
+    )
 }
 
-object NamedLens {
+object FieldLens {
+
+  type Id[A] = FieldLens[A, A]
+  val Id: FieldLens[Any, Any] = FieldLens(DataPath(Nil), identity[Any])
+  def id[A]: FieldLens[A, A] = Id.asInstanceOf[Id[A]]
+}
 
-  type Id[A] = NamedLens[A, A]
-  val Id: NamedLens[Any, Any] = NamedLens(DataPath(Nil), identity[Any])
-  def id[A]: NamedLens[A, A] = Id.asInstanceOf[Id[A]]
+// TODO: Move to enum-style inside NamedLens?
+final case class CompoundLens[A, B <: Product] private[CompoundLens] (
```

Comment:
This splitting of the lens into compound lens was a bit pre-mature. I will likely need the ability to select multiple values as a tuple, but I'm not sure packing it all into the Lens construct is the right way to do it. I might revert these changes for now.

A better approach might be to create SelectTuple{2-10} expressions that accept up to 10 different lens to produce a Tuple10 result. These can all extend from the same sealed trait so the visitor can just operate on the parent trait and abstract over arity as needed.

Either way, I think it is not needed for any upcoming work, so I should clean this up.